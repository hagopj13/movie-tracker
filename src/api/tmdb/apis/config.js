// @flow
import axios from '../axiosInstance';

export const getConfig = () => axios.get('/configuration');

export const getAllGenres = () => axios.get('/genre/movie/list');
