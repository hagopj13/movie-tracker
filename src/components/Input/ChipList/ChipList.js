// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Chip from '@material-ui/core/Chip';

type Props = {
  items: Array<{ id: number, name: string }>,
  selectedItems: number[],
  onItemClick: (id: number) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    margin: theme.spacing(-1, 0, 0, -1),
    '& > *': {
      margin: theme.spacing(1, 0, 0, 1),
    },
  },
  chipLabel: {
    fontSize: 12,
  },
  selectedChip: {
    color: theme.palette.common.white,
    border: `1px solid ${theme.palette.primary.main}`,
  },
}));

const ChipList = (props: Props) => {
  const { items, selectedItems, onItemClick } = props;

  const classes = useStyles();

  const isItemSelected = (id: number) => selectedItems.includes(id);

  return (
    <div className={classes.root}>
      {items.map((item) => (
        <Chip
          key={item.id}
          className={isItemSelected(item.id) ? classes.selectedChip : null}
          classes={{ label: classes.chipLabel }}
          label={item.name}
          color={isItemSelected(item.id) ? 'primary' : 'default'}
          variant={isItemSelected(item.id) ? 'default' : 'outlined'}
          onClick={() => onItemClick(item.id)}
        />
      ))}
    </div>
  );
};

export default ChipList;
