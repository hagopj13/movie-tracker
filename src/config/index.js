import React from 'react';
import DiscoverIcon from '@material-ui/icons/Movie';
import UpcomingIcon from '@material-ui/icons/Update';
import ProfileIcon from '@material-ui/icons/Person';

export const navigationItems = [
  {
    id: 'DISCOVER',
    label: 'Discover',
    to: '/',
    icon: <DiscoverIcon />,
    requiresAuth: false,
  },
  {
    id: 'UPCOMING',
    label: 'Upcoming',
    to: '/upcoming',
    icon: <UpcomingIcon />,
    requiresAuth: false,
  },
  {
    id: 'PROFILE',
    label: 'Profile',
    to: '/profile',
    icon: <ProfileIcon />,
    requiresAuth: true,
  },
];

export const dialogTypes = {
  LOGIN: 'LOGIN',
};

export const sortOptions = [
  { value: 'popularity.desc', label: 'Popularity Descending' },
  { value: 'popularity.asc', label: 'Popularity Ascending' },
  { value: 'vote_average.desc', label: 'Rating Descending' },
  { value: 'vote_average.asc', label: 'Rating Ascending' },
  { value: 'release_date.desc', label: 'Release Date Descending' },
  { value: 'release_date.asc', label: 'Release Date Ascending' },
];
