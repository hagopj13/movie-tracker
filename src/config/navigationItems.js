import React from 'react';
import DiscoverIcon from '@material-ui/icons/Movie';
import UpcomingIcon from '@material-ui/icons/Update';
import ProfileIcon from '@material-ui/icons/Person';

export default [
  {
    id: 'DISCOVER',
    label: 'Discover',
    to: '/',
    icon: <DiscoverIcon />,
    requiresAuth: false,
  },
  {
    id: 'UPCOMING',
    label: 'Upcoming',
    to: '/upcoming',
    icon: <UpcomingIcon />,
    requiresAuth: false,
  },
  {
    id: 'PROFILE',
    label: 'Profile',
    to: '/profile',
    icon: <ProfileIcon />,
    requiresAuth: true,
  },
];
