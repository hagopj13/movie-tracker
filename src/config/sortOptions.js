export default [
  { value: 'popularity.desc', label: 'Popularity Descending' },
  { value: 'popularity.asc', label: 'Popularity Ascending' },
  { value: 'vote_average.desc', label: 'Rating Descending' },
  { value: 'vote_average.asc', label: 'Rating Ascending' },
  { value: 'release_date.desc', label: 'Release Date Descending' },
  { value: 'release_date.asc', label: 'Release Date Ascending' },
];
