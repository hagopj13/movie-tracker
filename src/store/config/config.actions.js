// @flow
import type { ImagesConfig, Genre } from 'types';

import ConfigActionTypes from './config.types';

const fetchImagesConfig = () => ({
  type: ConfigActionTypes.FETCH_IMAGES_CONFIG,
});

const fetchImagesConfigStart = () => ({
  type: ConfigActionTypes.FETCH_IMAGES_CONFIG_START,
});

const fetchImagesConfigSuccess = (imagesConfig: ImagesConfig) => ({
  type: ConfigActionTypes.FETCH_IMAGES_CONFIG_SUCCESS,
  payload: { imagesConfig },
});

const fetchImagesConfigFailure = (error: string) => ({
  type: ConfigActionTypes.FETCH_IMAGES_CONFIG_FAILURE,
  payload: { error },
});

const fetchAllGenres = () => ({
  type: ConfigActionTypes.FETCH_ALL_GENRES,
});

const fetchAllGenresStart = () => ({
  type: ConfigActionTypes.FETCH_ALL_GENRES_START,
});

const fetchAllGenresSuccess = (genres: Genre[]) => ({
  type: ConfigActionTypes.FETCH_ALL_GENRES_SUCCESS,
  payload: { genres },
});

const fetchAllGenresFailure = (error: string) => ({
  type: ConfigActionTypes.FETCH_ALL_GENRES_FAILURE,
  payload: { error },
});

export default {
  fetchImagesConfig,
  fetchImagesConfigStart,
  fetchImagesConfigSuccess,
  fetchImagesConfigFailure,
  fetchAllGenres,
  fetchAllGenresStart,
  fetchAllGenresSuccess,
  fetchAllGenresFailure,
};
