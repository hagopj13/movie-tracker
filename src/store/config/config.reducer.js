// @flow
import { handleActions } from 'redux-actions';

import type { ImagesConfig, Genre } from 'types';

import ConfigActionTypes from './config.types';

type State = {
  images: ImagesConfig,
  genres: Genre[],
};

export const initialState: State = {
  images: {
    baseImageUrl: '',
    backdropSize: '',
    posterSize: '',
    profileSize: '',
  },
  genres: [],
};

const fetchImagesConfigSuccess = (state: State, action: any) => ({
  ...state,
  images: action.payload.imagesConfig,
});

const fetchAllGenresSuccess = (state: State, action: any) => ({
  ...state,
  genres: action.payload.genres,
});

const actionHandler = {
  [ConfigActionTypes.FETCH_IMAGES_CONFIG_SUCCESS]: fetchImagesConfigSuccess,
  [ConfigActionTypes.FETCH_ALL_GENRES_SUCCESS]: fetchAllGenresSuccess,
};

export default handleActions(actionHandler, initialState);
