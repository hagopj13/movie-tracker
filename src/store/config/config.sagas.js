import { takeLatest, all, call, put } from 'redux-saga/effects';

import * as api from 'api/tmdb';
import { convertResponseToImagesConfig } from 'api/tmdb/utils';

import ConfigActionTypes from './config.types';
import configActions from './config.actions';

export function* fetchImagesConfig() {
  yield put(configActions.fetchImagesConfigStart());
  try {
    const { data } = yield call(api.getConfig);
    const imagesConfig = yield call(convertResponseToImagesConfig, data);
    yield put(configActions.fetchImagesConfigSuccess(imagesConfig));
  } catch (error) {
    yield put(configActions.fetchImagesConfigFailure(error.status_message));
  }
}

export function* fetchAllGenres() {
  yield put(configActions.fetchAllGenresStart());
  try {
    const { data } = yield call(api.getAllGenres);
    yield put(configActions.fetchAllGenresSuccess(data.genres));
  } catch (error) {
    yield put(configActions.fetchAllGenresFailure(error.status_message));
  }
}

export function* onFetchImagesConfig() {
  yield takeLatest(ConfigActionTypes.FETCH_IMAGES_CONFIG, fetchImagesConfig);
}

export function* onFetchAllGenres() {
  yield takeLatest(ConfigActionTypes.FETCH_ALL_GENRES, fetchAllGenres);
}

export default function* configSagas() {
  yield all([call(onFetchImagesConfig), call(onFetchAllGenres)]);
}
