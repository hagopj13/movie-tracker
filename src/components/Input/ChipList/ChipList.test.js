import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import Chip from '@material-ui/core/Chip';

import { genres } from 'store/fixtures/config';

import ChipList from './ChipList';

describe('ChipList component', () => {
  let shallow;
  let items;
  let selectedItems;
  let wrapper;

  const mockItemClick = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    items = genres;
    selectedItems = [];
    const mockProps = {
      items,
      selectedItems,
      onItemClick: mockItemClick,
    };
    wrapper = shallow(<ChipList {...mockProps} />);
  });

  it('should render the ChipList component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Chip for each item', () => {
    expect(wrapper.find(Chip)).toHaveLength(items.length);
  });

  it('should set the Chip label to the item name', () => {
    expect(wrapper.find(Chip).at(0).prop('label')).toBe(items[0].name);
  });

  it('should call the onItemClick prop with the clicked item id', () => {
    wrapper.find(Chip).at(0).simulate('click');
    expect(mockItemClick).toHaveBeenLastCalledWith(items[0].id);
  });

  it('should set the primary color on the Chip if the item is selected', () => {
    selectedItems = [items[1].id];
    const mockProps = {
      items,
      selectedItems,
      onItemClick: mockItemClick,
    };
    const newWrapper = shallow(<ChipList {...mockProps} />);
    expect(newWrapper.find(Chip).at(1).prop('color')).toBe('primary');
    expect(newWrapper.find(Chip).at(0).prop('color')).toBe('default');
  });
});
