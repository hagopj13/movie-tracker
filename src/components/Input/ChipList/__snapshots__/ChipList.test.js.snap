// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipList component should render the ChipList component correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Chip))
    className={null}
    classes={
      Object {
        "label": "makeStyles-chipLabel-2",
      }
    }
    color="default"
    key="28"
    label="Action"
    onClick={[Function]}
    variant="outlined"
  />
  <WithStyles(ForwardRef(Chip))
    className={null}
    classes={
      Object {
        "label": "makeStyles-chipLabel-2",
      }
    }
    color="default"
    key="12"
    label="Adventure"
    onClick={[Function]}
    variant="outlined"
  />
  <WithStyles(ForwardRef(Chip))
    className={null}
    classes={
      Object {
        "label": "makeStyles-chipLabel-2",
      }
    }
    color="default"
    key="16"
    label="Animation"
    onClick={[Function]}
    variant="outlined"
  />
  <WithStyles(ForwardRef(Chip))
    className={null}
    classes={
      Object {
        "label": "makeStyles-chipLabel-2",
      }
    }
    color="default"
    key="35"
    label="Comedy"
    onClick={[Function]}
    variant="outlined"
  />
</div>
`;
